google/cloud/internal/future_void.h

```diff
@@ -0,0 +1,163 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_FUTURE_VOID_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_FUTURE_VOID_H_
+/**
+ * @file
+ *
+ * Fully specialize `future<void>` and `promise<R>` for void.
+ */
+
+#include "google/cloud/internal/port_platform.h"
+
+// C++ futures only make sense when exceptions are enabled.
+#if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+#include "google/cloud/internal/future_fwd.h"
+#include "google/cloud/internal/future_impl.h"
+
+namespace google {
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+/**
+ * Implement ISO/IEC TS 19571:2016 future for void.
+ */
+template <>
+class future<void> {
+ public:
+  future() noexcept : shared_state_() {}
+  future(future<void>&& f) noexcept
+      : shared_state_(std::move(f.shared_state_)) {}
+
+  // TODO(#1345) - implement the unwrapping constructor.
+  // future(future<future<void>>&& f) noexcept;
+  // future& operator=(future<future<void>>&& f) noexcept;
+  future(future<void> const&) = delete;
+
+  future& operator=(future<void>&& f) noexcept {
+    future tmp(std::move(f));
+    std::swap(tmp.shared_state_, shared_state_);
+    return *this;
+  }
+  future& operator=(future const&) = delete;
+
+  void get() {
+    check_valid();
+    std::shared_ptr<shared_state_type> tmp;
+    tmp.swap(shared_state_);
+    return tmp->get();
```

Comment:
return from void?

Reply:
It works because `tmp->get()` also returns void. The generic code (for `T`, not for `void`) also looks like this.